iree/compiler/Conversion/LinalgToVector/VectorizeConv.cpp

```diff
@@ -209,6 +209,125 @@ struct VectorizeLinalgConv : OpRewritePattern<linalg::ConvOp> {
+///   - Hf must be 1.
+/// - No dilation.
+/// - No padding.
+///
+/// Channel is requried to be a multiple of 4 so that we can process them with
+/// load4/store4, which is native to GPUs.
+struct VectorizeLinalgDepthwiseConv
+    : OpRewritePattern<linalg::DepthwiseConvInputNHWCFilterHWCOp> {
+  using OpRewritePattern::OpRewritePattern;
+
+  LogicalResult matchAndRewrite(
+      linalg::DepthwiseConvInputNHWCFilterHWCOp convOp,
+      PatternRewriter &rewriter) const override {
+    LLVM_DEBUG(llvm::dbgs() << "inspecting " << convOp << "\n");
+
+    auto inputViewOp = convOp.getInput(0).getDefiningOp<SubViewOp>();
+    auto filterViewOp = convOp.getInput(1).getDefiningOp<SubViewOp>();
+    auto outputViewOp = convOp.getOutput(0).getDefiningOp<SubViewOp>();
+    if (!filterViewOp || !inputViewOp || !outputViewOp) return failure();
+
+    // The filter/input/output view should have static sizes to vectorize.
+    if (!llvm::empty(filterViewOp.getDynamicSizes()) ||
+        !llvm::empty(inputViewOp.getDynamicSizes()) ||
+        !llvm::empty(outputViewOp.getDynamicSizes())) {
+      return failure();
+    }
+
+    // The output batch and height dimensions should be 1. If not, other
+    // patterns can generate parallel loops can distribute them.
+    if (outputViewOp.getStaticSize(0) != 1 ||
+        outputViewOp.getStaticSize(1) != 1) {
+      return failure();
+    }
+
+    // We addtionally expect the filter height/width dimensions are both 1 to
+    // simplify vectorization. Other patterns can generate loops to create 1x1
+    // filter subivews.
+    if (filterViewOp.getStaticSize(0) != 1 ||
+        filterViewOp.getStaticSize(1) != 1) {
+      return failure();
+    }
+
+    int64_t numChannels = outputViewOp.getStaticSize(3);
+    if (numChannels % 4 != 0) return failure();
+
+    int64_t numOutputWidths = outputViewOp.getStaticSize(2);
+    int64_t widthStride = convOp.strides().getValue<int64_t>({1});
+
+    // This invocation handles a batch of (numOutputWidths * numOutputChannels).
+    LLVM_DEBUG({
+      llvm::dbgs() << "# output width: " << numOutputWidths << "\n";
+      llvm::dbgs() << "# channels: " << numChannels << "\n";
+      llvm::dbgs() << "width stride: " << widthStride << "\n";
+    });
+
+    Location loc = convOp.getLoc();
+
+    Type elementType = filterViewOp.getType().getElementType();
+    auto vector4Type = VectorType::get(4, elementType);
+    auto filterVectorType = VectorType::get({numChannels}, elementType);
+    Value zero = rewriter.createOrFold<ConstantIndexOp>(loc, 0);
+
+    // Load the entire filter subview.
+    SmallVector<Value, 4> filterIndices(3, zero);
+    Value wholeFilter = rewriter.create<vector::TransferReadOp>(
+        loc, filterVectorType, filterViewOp, filterIndices);
+
+    // Compute the (numOutputWidths * numOutputChannels) output batch. We only
+    // contribute numChannels accumulation along the reduction dimension.
+    for (int oc = 0; oc < numChannels / 4; ++oc) {
```

Comment:
Can we create a bigger vector here and let unroll break it down?

Reply:
Not sure I get your idea here properly. But this is already doing whole vector transfer read (L291) and using extract strided slice to take slices of it (L297). Vector unroll will break it down later. 

Or do you mean to put the output channel as the innermost to process it as a whole so that vector unroll can break it down? I'd like to read in the filter first so that it can be shared in the loops for output height/width though.